File: CS/Form1.cs
using System;
using System.Drawing;
using System.Drawing.Imaging;
using System.IO;
using System.Windows.Forms;
using DevExpress.Office.Services;

namespace RichIUriStreamProviderExample {
    public partial class Form1 : Form {
        private string basePath;
        
        public Form1() {
            InitializeComponent();

            basePath = Directory.GetCurrentDirectory() + @"\..\..\TestDocs\";

            // Custom IUriStreamProvider registration
            IUriStreamService uriStreamService = richEditControl1.GetService<IUriStreamService>();
            uriStreamService.RegisterProvider(new CustomUriStreamProvider(basePath, "bmp"));
        }

        private void simpleButton1_Click(object sender, EventArgs e)
        {
            richEditControl1.LoadDocument(basePath + "test.html");
        }
    }

    public class CustomUriStreamProvider : IUriStreamProvider {
        private string basePath;
        private string imageExtension;
        public string BasePath { get { return basePath; } set { basePath = value; } }
        public string ImageExtension { get { return imageExtension; } set { imageExtension = value; } }

        public CustomUriStreamProvider(string basePath, string imageExtenssion) {
            BasePath = basePath;
            ImageExtension = imageExtenssion;
        }

        public Stream GetStream(string url) {
            string fileName = string.Format("{0}.{1}", url.Replace("cid:", string.Empty), ImageExtension);
            MemoryStream memoryStream = new MemoryStream();
            Image image = Image.FromFile(BasePath + fileName);

            image.Save(memoryStream, ImageFormat.Bmp);
            memoryStream.Seek(0, SeekOrigin.Begin);

            return memoryStream;
        }
    }
}

File: VB/Form1.vb
Imports System
Imports System.Drawing
Imports System.Drawing.Imaging
Imports System.IO
Imports System.Windows.Forms
Imports DevExpress.Office.Services

Namespace RichIUriStreamProviderExample
    Partial Public Class Form1
        Inherits DevExpress.XtraEditors.XtraForm

        Private basePath As String

        Public Sub New()
            InitializeComponent()

            basePath = Directory.GetCurrentDirectory() & "\..\..\TestDocs\"

            ' Custom IUriStreamProvider registration
            Dim uriStreamService As IUriStreamService = richEditControl1.GetService(Of IUriStreamService)()
            uriStreamService.RegisterProvider(New CustomUriStreamProvider(basePath, "bmp"))
        End Sub

        Private Sub btnImportFile_Click(ByVal sender As Object, ByVal e As EventArgs) Handles btnImportFile.Click
            richEditControl1.LoadDocument(basePath & "test.html")
        End Sub
    End Class

    Public Class CustomUriStreamProvider
        Implements IUriStreamProvider

        Private basePath_Renamed As String

        Private imageExtension_Renamed As String
        Public Property BasePath() As String
            Get
                Return basePath_Renamed
            End Get
            Set(ByVal value As String)
                basePath_Renamed = value
            End Set
        End Property
        Public Property ImageExtension() As String
            Get
                Return imageExtension_Renamed
            End Get
            Set(ByVal value As String)
                imageExtension_Renamed = value
            End Set
        End Property

        Public Sub New(ByVal basePath As String, ByVal imageExtenssion As String)
            Me.BasePath = basePath
            ImageExtension = imageExtenssion
        End Sub

        Public Function GetStream(ByVal url As String) As Stream Implements IUriStreamProvider.GetStream
            Dim fileName As String = String.Format("{0}.{1}", url.Replace("cid:", String.Empty), ImageExtension)
            Dim memoryStream As New MemoryStream()
            Using image As Image = System.Drawing.Image.FromFile(BasePath & fileName)
                image.Save(memoryStream, ImageFormat.Bmp)
            End Using
            memoryStream.Seek(0, SeekOrigin.Begin)
            Return memoryStream
        End Function
    End Class
End Namespace